.circleci/config.yml

```diff
@@ -1017,31 +1017,74 @@ jobs:
-              ~/datadog/zend_abstract_interface
-            make -j all
-            make test
-
-      - run:
-          name: Build and test Zend Abstract Interface (UBSan)
+      # Builds that support ASan
+      - when:
+          condition:
+            or:
+              - equal: [ "7.4", << parameters.php_version >> ]
+              - equal: [ "8.0", << parameters.php_version >> ]
+          steps:
+            - run:
+                name: Build and test Zend Abstract Interface (nts, debug, debug-zts-asan)
+                command: |
+                  export PATH="/opt/cmake/<< parameters.cmake_version >>/bin:$PATH"
+                  for phpBuild in nts debug debug-zts-asan
+                  do
+                    switch-php ${phpBuild}
+                    mkdir -p /tmp/build/zai-${phpBuild}
+                    cd /tmp/build/zai-${phpBuild}
+                    toolchain=""
+                    if [ "${phpBuild}" = "debug-zts-asan" ]
+                    then
+                      toolchain="-DCMAKE_TOOLCHAIN_FILE=~/datadog/cmake/asan.cmake"
+                    fi
+                    CMAKE_PREFIX_PATH=/opt/catch2 cmake \
+                      ${toolchain} \
+                      -DCMAKE_BUILD_TYPE=Debug \
+                      -DBUILD_ZAI_TESTING=ON \
+                      -DPHP_CONFIG=$(which php-config) \
+                      ~/datadog/zend_abstract_interface
+                    make -j all
+                    make test
+                  done
+
+      # Builds that do not support ASan
+      - when:
+          condition:
+            or:
+              - equal: [ "5.4", << parameters.php_version >> ]
+              - equal: [ "5.5", << parameters.php_version >> ]
+              - equal: [ "5.6", << parameters.php_version >> ]
+              - equal: [ "7.0", << parameters.php_version >> ]
+              - equal: [ "7.1", << parameters.php_version >> ]
+              - equal: [ "7.2", << parameters.php_version >> ]
+              - equal: [ "7.3", << parameters.php_version >> ]
+          steps:
+            - run:
+                name: Build and test Zend Abstract Interface (nts, debug, debug-zts)
+                command: |
+                  export PATH="/opt/cmake/<< parameters.cmake_version >>/bin:$PATH"
+                  for phpBuild in nts debug debug-zts
+                  do
+                    switch-php ${phpBuild}
+                    mkdir -p /tmp/build/zai-${phpBuild}
+                    cd /tmp/build/zai-${phpBuild}
+                    CMAKE_PREFIX_PATH=/opt/catch2 cmake \
+                      -DCMAKE_BUILD_TYPE=Debug \
+                      -DBUILD_ZAI_TESTING=ON \
+                      -DPHP_CONFIG=$(which php-config) \
+                      ~/datadog/zend_abstract_interface
+                    make -j all
+                    make test
+                  done
+
+      - run:
+          name: Build and test Zend Abstract Interface (debug + UBSan)
```

Comment:
ubsan runs on all PHP versions?

Reply:
It does, but since we don't have UBSan builds of PHP, I don't think UBSan will flag anything on the PHP side of things so it should only catch issues with the ZAI components.